Swagger API for Azure CDN WebApplicationFirewallPolicy resource

sushantsingh opened the pull request:
This PR includes the following changes

Adds
1. Added the swagger api spec for CdnWebApplicationFirewallPolicy which is a new (public preview) arm resource under the Microsoft.Cdn provider
2. Added the relevant examples 

Edits
1. To integrate this with Azure CDN, a new property "webApplicationFirewallPolicyLink" is added to the "Endpoint" sub resource in cdn.json arm resource.  This new property is the means by which a "CdnWebApplicationFirewallPolicy" can be references in a cdn "Endpoint" resource.
2. All the examples related to the "Endpoint" resource in cdn.json were edited to include this new property
3. Readme.md, a new package was added for this new change


### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

sushantsingh committed:
Create cdnwebapplicationfirewall.json

sushantsingh committed:
Create CdnWafCreateOrUpdate

sushantsingh committed:
Create CdnWafListPolicies.json

sushantsingh committed:
Create CdnWafPolicyGet.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Create WafListManagedRuleSets.json

sushantsingh committed:
Create WafListPolicies.json

sushantsingh committed:
Create WafPolicyGet.json

sushantsingh committed:
Update WafListPolicies.json

sushantsingh committed:
Create WafPolicyCreateOrUpdate.json

sushantsingh committed:
Create WafPolicyDelete.json

sushantsingh committed:
Delete CdnWafCreateOrUpdate

sushantsingh committed:
Delete CdnWafListPolicies.json

sushantsingh committed:
Delete CdnWafPolicyGet.json

sushantsingh committed:
Create cdn.json

sushantsingh committed:
Create readme.md

sushantsingh committed:
Update cdn.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update cdn.json

sushantsingh committed:
Update WafListManagedRuleSets.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Create Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Get.json

sushantsingh committed:
Create Endpoints_Get.json

sushantsingh committed:
Update Endpoints_Get.json

sushantsingh committed:
Create Endpoints_Update.json

sushantsingh committed:
Update cdn.json

sushantsingh committed:
Added examples which shows default behaviour

sushantsingh committed:
Update Endpoints_Get.json

sushantsingh committed:
Update Endpoints_Update.json

sushantsingh committed:
Update Endpoints_ListByProfile.json

sushantsingh committed:
Update Endpoints_Start.json

sushantsingh committed:
Update Endpoints_Stop.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Delete cdn.json

sushantsingh committed:
Delete Endpoints_Create.json

sushantsingh committed:
Delete Endpoints_Get.json

sushantsingh committed:
Delete Endpoints_Update.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

sushantsingh committed:
Update cdn.json

sushantsingh committed:
Update Endpoints_Get.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_ListByProfile.json

sushantsingh committed:
Update Endpoints_Update.json

sushantsingh committed:
Update Endpoints_Start.json

sushantsingh committed:
Update Endpoints_Stop.json

sushantsingh committed:
Update cdn.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Get.json

sushantsingh committed:
Update Endpoints_ListByProfile.json

sushantsingh committed:
Update Endpoints_Start.json

sushantsingh committed:
Update Endpoints_Stop.json

sushantsingh committed:
Update Endpoints_Update.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update Endpoints_Create.json

sushantsingh committed:
Update WafListPolicies.json

sushantsingh committed:
Update WafListManagedRuleSets.json

sushantsingh committed:
Update WafPolicyCreateOrUpdate.json

sushantsingh committed:
Update WafPolicyGet.json

sushantsingh committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

sushantsingh committed:
Merge branch 'master' of https://github.com/sushantsingh/azure-rest-api-specs

sushantsingh committed:
Added new cdn version 2016-06-15

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-ruby

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-go

sushantsingh committed:
Update WafListPolicies.json

sushantsingh committed:
Update WafPolicyGet.json

sushantsingh committed:
Delete readme.md

sushantsingh committed:
Update readme.md

sushantsingh committed:
Update readme.md

sushantsingh committed:
Fix model validation failures

sushantsingh committed:
Merge branch 'master' of https://github.com/sushantsingh/azure-rest-api-specs

sushantsingh committed:
Update readme.md

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update WafPolicyCreateOrUpdate.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update readme.go.md

AutorestCI commented:
# Automation for azure-sdk-for-java

adxsdknet commented:
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/6183 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/6183/commits/52a697e26d4650fb6302c657519aec14c5033c29

nschonni commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+          "type": "string"
+        },
+        "anomalyScore": {
+          "description": "Verizon only : If the rule set supports anomaly detection mode, this descibes the threshold for blocking requests.",
```
```suggestion
          "description": "Verizon only : If the rule set supports anomaly detection mode, this describes the threshold for blocking requests.",
```

nschonni committed:
Update specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15/cdnwebapplicationfirewall.json

Co-Authored-By: sushantsingh <sushant.singh1987@gmail.com>

veronicagg reviewed (changes requested):
@sushantsingh if you're going to copy an existing API version an edit a new one, you can use [OpenAPI Hub](http://aka.ms/openapihub) and choose "Create OpenAPI specs". 
Additionally, please take a look at CI issues at https://dev.azure.com/azure-sdk/public/_build/results?buildId=28466&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=69&lineEnd=70&colStart=1&colEnd=1 and https://dev.azure.com/azure-sdk/public/_build/results?buildId=28466&view=logs&jobId=688669d0-441c-57c3-cf6d-f89a22ccfa5d&taskId=923b5a90-e462-5994-c166-86509c6bf5f2&lineStart=2315&lineEnd=2316&colStart=1&colEnd=1

sushantsingh commented:
@veronicagg I am working on the LintDiff failures. However the Model Validation failures are unrelated to my change. It looks like they have always existed for cdn.json. I have only added a single property to cdn.json. The crux of my change is adding a new resource cdnwebapplicationpolicies.json.
How do I suppress the unrelated Model Validation failures? Ideally there should be something similar to LintDiff failures that detects what new failures were introduced.

We did the swagger review with a couple of folks from the arm api review board, should we schedule one with you as well to help you understand the changes better?

sushantsingh committed:
add int max mins where they make sense

sushantsingh committed:
Update WafPolicyDelete.json

sushantsingh committed:
Added Patch for tags

sushantsingh committed:
Provisioning state as enum

sushantsingh committed:
Add preview suffix to the version

sushantsingh committed:
Update readme.md

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Create WafPatchPolicy.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh commented:
@veronicagg do you need anything else from my side for this pull request? All the validations are passing except ModelValidations which are old unrelated errors in cdn.json which have always been there. I have just added a single property to cdn.json and there are no failures related to that. I updated the pull request description to explain that better.

dsgouda commented:
startbuild

adxsdknet commented:
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/6183 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/6183/commits/904da172ce557aa99fb564b33bc18e37f7baf65c

veronicagg reviewed (changes requested):
@sushantsingh thanks for updating the description with comments. New api versions will report all issues from tools, as the recommendation is to fix them all, if you're not planning on fixing those, we can still go ahead and they will continue to be reported. The PR is hard to review as is, since there are so many commits, we ask for folks to generally commit their previous api version and make the changes in new commits (short list). If you're already reviewed with ARM, could you ask them to sign-off on the PR? I added the label a few days back, but haven't seen activity from them here.

veronicagg commented on specification/cdn/resource-manager/readme.md:
```diff
@@ -25,8 +25,25 @@ To see additional help and options, run:
+directive:
+  - where:
+      - $.paths
+    suppress:
```
was this exception approved by ARM team?

veronicagg commented on specification/cdn/resource-manager/readme.md:
````diff
@@ -25,8 +25,25 @@ To see additional help and options, run:
+
+``` yaml $(tag) == 'package-2019-06'
+input-file:
+- Microsoft.Cdn/stable/2019-06-15/cdn.json
````
are you intentionally combining a stable and preview version of APIs in your latest package? This is the package that will be used to ship SDKs from by default.

veronicagg reviewed (changes requested):
couple of inline questions, rest looks good to me. thanks@

sushantsingh commented:
@veronicagg I have sent an email to the two reviewers from the ARM api board (David Justice and Marcin Jastrzebski) to see if they can sign off. You're on that thread as well.

sushantsingh committed:
Moved everything to the preview version

sushantsingh committed:
Update readme.md

veronicagg commented:
@sushantsingh do you have any questions on the job failures? looks like some file paths are incorrect, see https://dev.azure.com/azure-sdk/public/_build/results?buildId=30680&view=logs&jobId=95dbed5b-845d-57ba-33d3-f6c5e69afcf4&taskId=44f1510e-a822-56b0-2e0b-241e0a387c6a&lineStart=74&lineEnd=75&colStart=1&colEnd=1

veronicagg commented:
@sushantsingh I'm changing roles, so @amarzavery will take over this review. thanks!

amarzavery commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+        },
+        "sku": {
+          "description": "The pricing tier (defines a CDN provider, feature list and rate) of the CdnWebApplicationFirewallPolicy.",
+          "$ref": "./network.json#/definitions/Sku"
```
The file "./network.json" does not exist. This should probably be `"./cdn.json#/definitions/Sku"` as I can see that definition in cnd.json over [here](https://github.com/Azure/azure-rest-api-specs/pull/5885/files#diff-6d53702dc9b0abd52dedefbde8482955R4272).

amarzavery replied:
ping @sushantsing !!

amarzavery replied:
ping @sushantsingh  !!

sanjaiganesh commented:
@sushantsingh Could you please add old files first (new iteration) and then paste new content (another iteration) so that I can see the difference.. it is not practical to comb through these huge files to figure out what changed. thanks.

nschonni commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+        "tags": [
+          "WebApplicationFirewallPolicies"
+        ],
+        "description": "Update an existing CdnWebApplicationFirewallPolicy with the specified policy name uner the specified subcription and resource group",
```
```suggestion
        "description": "Update an existing CdnWebApplicationFirewallPolicy with the specified policy name under the specified subscription and resource group",
```

nschonni commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "description": "CdnWebAplicationFirewallPolicy parameters to be patched.",
```
```suggestion
            "description": "CdnWebApplicationFirewallPolicy parameters to be patched.",
```

devigned replied:
@sushantsingh please fix this.

nschonni replied:
There is a spellcheck job now (although it was failing for a bit because of an issue in another API, so it was easy to ignore 😄 ) https://dev.azure.com/azure-sdk/public/_build/results?buildId=44986&view=logs&jobId=eca5a1cb-f12a-57d8-9254-390bd587c201

nschonni reviewed (commented):
minor typos

sushantsingh committed:
Update WafPolicyCreateOrUpdate.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update Profiles_Delete.json

sushantsingh committed:
Update Endpoints_Delete.json

sushantsingh committed:
Update Endpoints_PurgeContent.json

sushantsingh committed:
Update Endpoints_LoadContent.json

sushantsingh committed:
Update CustomDomains_Delete.json

sushantsingh committed:
Update CustomDomains_DisableCustomHttps.json

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingCDNManagedCertificate.json

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingBYOC.json

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh committed:
Update readme.go.md

sushantsingh committed:
Update cdnwebapplicationfirewall.json

KrisBash commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cdn/CdnWebApplicationFirewallPolicies/{policyName}": {
```
What's the relationship between a cdnWebApplicationFirewallPolicies resource type and a [Microsoft.Network/webApplicationFirewallPolicies ](https://github.com/Azure/azure-rest-api-specs/blob/7816a440b356998df165e0dd84aa368794c22028/specification/network/resource-manager/Microsoft.Network/stable/2018-12-01/webapplicationfirewall.json )resource ?  Is there an opportunity to use an extension resource, such as {resource URI}/providers/Microsoft.Network/webApplicationFirewallsPolicies/{resource name} in lieu of a new duplicate resource type? If not, does the schema deviate from the .Network variant?

sushantsingh replied:
This is not a duplicate, it is for Azure Cdn product (which works with multiple third party providers like Akamai and Verizon) and has a different schema and is expected to evolve seperately.

KrisBash commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+        },
+        "x-ms-long-running-operation": true
+      },
+      "patch": {
```
While PATCH is only mandatory for tags, we prefer to see PATCH support for all modifiable properties.

devigned replied:
@sushantsingh any response to this?

sushantsingh replied:
We will do this in the stable version of the resource. As of now, as recommended in the Arm API review meeting, the minimum bar is to be able to update tags and that has been included in the API.
We do not have any scenario where we update parts of the config, the entire config is updated as a single entity.

KrisBash commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+        "x-ms-long-running-operation": false
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Cdn/CdnWebApplicationFirewallManagedRuleSets": {
```
Can you explain what this resource is? It is not clear to me it's relationship with the other new resources.

sushantsingh replied:
I have already had the ARM review in person with the Arm API board where I explained all the context etc. about this resource.
It would be very inefficient to do it all over again especially via comments.

devigned replied:
@sushantsingh in cases where you would not like to explain again, if you can point to a design doc or something that explains a bit more, that would be super helpful.

sushantsingh replied:
Sure I can do that for the future pull requests, for now I do not have this document and I am working hard to get the service ready for the release date.

sushantsingh replied:
The jist of the change is included in the pull request description.

KrisBash commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+{
+  "swagger": "2.0",
```
A subscription level list API is missing, i.e. `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup Name}/providers/Microsoft.CDN/cdnWebApplicationFirewallPolicies

devigned replied:
@sushantsingh please respond.

sushantsingh replied:
We do have this API in the swagger, please clarify if I am missing something

`  "paths": {
    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cdn/CdnWebApplicationFirewallPolicies": {
      "get": {
        "tags": [
          "WebApplicationFirewallPolicies"
        ],
        "description": "Lists all of the protection policies within a resource group.",
        "operationId": "Policies_List",
        "parameters": [
          {
            "$ref": "#/parameters/ResourceGroupNameParameter"
          },
          {
            "$ref": "#/parameters/SubscriptionIdParameter"
          },
          {
            "$ref": "#/parameters/ApiVersionParameter"
          }
        ],
        "responses": {
          "200": {
            "description": "OK. The request has succeeded.",
            "schema": {
              "$ref": "#/definitions/CdnWebApplicationFirewallPolicyList"
            }
          },
          "default": {
            "description": "Error response describing why the operation failed.",
            "schema": {
              "$ref": "./cdn.json#/definitions/ErrorResponse"
            }
          }
        },
        "x-ms-pageable": {
          "nextLinkName": "nextLink"
        },
        "x-ms-examples": {
          "List Policies in a Resource Group": {
            "$ref": "./examples/WafListPolicies.json"
          }
        }
      }
    },`

sushantsingh replied:
We will work on this API for GA.

KrisBash reviewed (changes requested):
Please see questions in comments

devigned commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+      "properties": {
+        "rateLimitThreshold": {
+          "description": "Defines rate limit threshold.",
+          "type": "integer"
```
Does this take negatives? Is there a maximum?

sushantsingh replied:
No, it cannot take negatives, I added a minimum constraint, thanks for catching this!
There is no maximum in my knowledge at the moment, it will just be the int max for now.

devigned commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+      "properties": {
+        "ruleId": {
+          "description": "Identifier for the managed rule.",
+          "type": "string"
```
Does it have any constraints?

sushantsingh replied:
this will be discovered via the GET managed rule api and the same name is expected to be used here. In the future when we release the "Stable" version, we will do some work to convert this into an enum with fixed values, but for the preview version, since new managedRuleSets come up  often we do not want to create barriers for their adoption.
The check for unsupported ruleSets for the "preview" version will be on the RP service.

devigned replied:
ok for preview

devigned commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+    "ManagedRuleDefinition": {
+      "description": "Describes a managed rule definition.",
+      "properties": {
+        "ruleId": {
```
Same comment as previous ruleId. Should it have any constraints?

sushantsingh replied:
This is all discovered via GET ManagedRuleSet API and when we have more clients like PS, CLI etc. customers will directly plumb things from the GET to the PUT. We will look into adding more constraints to the swagger in the "stable" version once we have a better understanding of the constraints. Things are in active development right now and are subject to change.

devigned commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdn.json:
```diff
@@ -0,0 +1,4353 @@
+          }
+        ],
+        "responses": {
+          "200"
```
This should be a duration, not a string.

sushantsingh replied:
@devigned I am not sure what you meant on here, can you please clarify if posible.
Also, this is completely unrelated to me, all I have done is copied the cdn.json from the stable folder and added one line to the swagger which was needed to add a link from cdn endpoint to the cdnWebApplicationFirewallPolicies. You can find that line by searching for "cdnWebApplicationFirewallPolicies"

devigned replied:
This was meant for cache duration: https://github.com/Azure/azure-rest-api-specs/pull/5885/files/7f720a74e69ddb58977ad020865a53d5f24a06cf#diff-6d53702dc9b0abd52dedefbde8482955R3372. I'm not sure why the line was marked at 1948.

devigned replied:
The cacheDuration should be https://en.wikipedia.org/wiki/ISO_8601#Durations

sushantsingh replied:
@devigned ok got it. In this case, this is what is in the stable cdn version 
https://github.com/sushantsingh/azure-rest-api-specs/blob/master/specification/cdn/resource-manager/Microsoft.Cdn/stable/2019-04-15/cdn.json

My only change to cdn.json (other than copying it to preview folder) is
        "webApplicationFirewallPolicyLink": {
          "description": "Defines the Web Application Firewall policy for the endpoint (if applicable)",
          "type": "object",
          "properties": {
            "id": {
              "type": "string",
              "description": "Resource ID."
            }
          }
        }

devigned replied:
I understand, but that doesn't change that it should be a ISO 8601 duration.

sushantsingh replied:
I meant to say above that it is part of the existing cdn.json swagger and is not introduced by me. The reason it shows up in the diff is because the one change I did make to the cdn swagger needs to be a preview change as well, so I had to copy the entire thing into the preview folder.

sushantsingh replied:
Will start a email thread with David Justice and the owner of cdn.json

devigned reviewed (changes requested):
I have some questions and some actionable feedback. Please address or respond to these items.

sushantsingh commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cdn/CdnWebApplicationFirewallPolicies": {
```
no there is no scope of doing this at the moment. It is a management decision and is consistent with other waf producsts for appgateway and frontdoor (frontdoorwebapplicaitonfirewallpolicies)

sushantsingh commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,919 @@
+      "properties": {
+        "name": {
+          "description": "Defines the name of the custom rule",
+          "type": "string"
```
@devigned as we discussed in the Api review meeting, the name is consistent with afd's waf which is frontdoorwebapplicationfirewallpolicies, we deliberated on this internally and decided to go with CdnWebApplicationFirewallPolicies instead of cdnwafpolicies.

As far as the rule name is concerned, it is arbitrary at the moment, we will look into adding more constraints as part of the work for the "Stable" version.

devigned replied:
So I can provide you a 10000 char string?

devigned replied:
please add constraints in stable

sushantsingh committed:
Update cdnwebapplicationfirewall.json

sushantsingh commented:
Hi @devigned  do you have any other pending concerns?

devigned commented:
@sushantsingh please resolve the build issue. I know it is from a previous example, but it's not a good reason to keep propagating incorrect examples. Thank you!

sushantsingh commented:
@devigned Sure I can try to fix the build issue but I have no idea what it means (I did try my best to reduce the error count due to existing issues by fixing everything I could understand).
Can you please help me understand what "message: Cannot read property 'readUInt8' of null" means? Please expedite this if possible, our release date of June 15th.

Validating "examples" and "x-ms-examples" in  specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdn.json:

 error : ‌
operationId: CustomDomains_EnableCustomHttps
scenario: CustomDomains_EnableCustomHttpsUsingYourOwnCertificate
source: request
responseCode: ALL
severity: 0
code: INTERNAL_ERROR
details:
  **message: Cannot read property 'readUInt8' of null**
  code: INTERNAL_ERROR

sushantsingh committed:
Update specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json

Co-Authored-By: Nick Schonning <nschonni@gmail.com>

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingCDNManagedCertificate.json

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingBYOC.json

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingBYOC.json

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingCDNManagedCertificate.json

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingBYOC.json

sushantsingh committed:
Update CustomDomains_EnableCustomHttpsUsingCDNManagedCertificate.json

nschonni commented on specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json:
```diff
@@ -0,0 +1,921 @@
+        "tags": [
+          "WebApplicationFirewallPolicies"
+        ],
+        "description": "Update an existing CdnWebApplicationFirewallPolicy with the specified policy name under the specified subcription and resource group",
```
```suggestion
        "description": "Update an existing CdnWebApplicationFirewallPolicy with the specified policy name under the specified subscription and resource group",
```

sushantsingh committed:
Update specification/cdn/resource-manager/Microsoft.Cdn/preview/2019-06-15-preview/cdnwebapplicationfirewall.json

Co-Authored-By: Nick Schonning <nschonni@gmail.com>

devigned reviewed (approved):
Approved with the understanding that subscription level get will be added and cacheDuration will be converted to a ISO duration rather than a string before a stable release.

devigned commented:
@KrisBash are you good for preview. @sushantsingh and I chatted and we have commitment to add a subscription level get, improved documentation for models / operations, and to start a conversation about migrating cache duration to an ISO duration rather than string for the stable release.

sushantsingh commented:
Hi folk, this needs to be merged soon or we risk not meeting the preview release date of Jun 15th. Please let me know if anything is needed from our end to expedite things.
@KrisBash @amarzavery @devigned

KrisBash reviewed (approved):
Signing off with understanding that subscription level GET will be added for stable API version
